### DISCUSSION TITLE
GPU memory usage not consistent with estimated usage

### DISCUSSION THREAD

**Discussion by liuchihl** [CliMA/Oceananigans.jl#3557] [Helpdesk]
Hi,

I've conducted a few 3D experiments to examine the relationship between actual GPU memory usage and estimated GPU memory usage. The estimated GPU memory usage is calculated using the formula: `Nx x Ny x Nz x 8 x N_state`, where `Nx, Ny, Nz` are the grid dimensions, the factor of 8 represents 64-bit floating-point numbers, and `N_state` is the number of state variables. Since buoyancy is the only tracer, the state variables are `u, v, w, b, p, du/dt, dv/dt, dw/dt, db/dt`.

The figure below illustrates that the GPU usage can be scaled linearly with my estimation, which is great. However, the actual GPU memory usage is approximately 3.7 times larger than my initial estimate (y-axis). This suggests that there may be additional state variables that were not accounted for in my estimation. I'm interested in identifying any overlooked variables. Alternatively, if my estimation is correct, I would like to understand the reasons for this discrepancy. Any thoughts would be appreciated!

More information that might help:
1. running longer does not seem to change the GPU memory usage, indicating there is no memory leak
2. The above experiments include 12 variables that are written in netcdf output writer. But writing 4 variables (another experiment) as output is only 14% less than writing 12 variables. Therefore, I think the number of variables does not explain why there is a discrepancy between estimated GPU memory usage and actual GPU usage. 
3. I am running on Oceananigans@v0.90.11 with Julia 1.10.2.

Chih-Lun
![benchmark_mem](https://github.com/CliMA/Oceananigans.jl/assets/68127124/a5e93415-66ff-4900-88df-0204d924db95)

**Comment by glwagner** [comment]
Each prognostic variable has the cost of 3 fields --- 1 for the state, and 2 for the tendencies (it looks like you're only allowing for one tendency term per field). You may also want to account for halo regions (so each field has size `Nx + 2Hx`, etc). For the `NonhydrostaticModel` there are two pressure fields (one for hydrostatic pressure, one for nonhydrostatic pressure anomaly). The closures sometimes allocate fields. Also, the nonhydrostatic pressure solver allocates at least one additional complex-valued field for scratch memory (I think there can occasionally be a second, depending on the topology and also whether its on GPU vs CPU --- take a look at the `FFTBasedPoissonSolver`).

Let's see how far off the estimate is after accounting for all that...

**Reply by liuchihl** [reply]
Thank you for these suggestions! I'll keep you updated on the progress of the estimation.

**Reply by liuchihl** [reply]
I have now estimated the following variables: u, v, w, b, du/dt, du*/dt, dv/dt, dv*/dt, dw/dt, dw*/dt, db/dt, db*/dt, p, p', p_complex, and closure (which is defined as ScalarDiffusivity(; ν=1e-4, κ=1e-4), but I am uncertain about the number of fields that are allocated), give or take N_state = 16. Considering the halo points and excluding other diagnostics, and keeping only u, v, w, and b as my output, I get:
![benchmark_mem_better_estimation](https://github.com/CliMA/Oceananigans.jl/assets/68127124/b0aead0b-870a-48cf-b369-4230e027c7aa)

I am glad it is getting closer to the 1 to 1 line, but there is still a slight discrepancy, I might have overlooked some other factors.

**Reply by glwagner** [reply]
`ScalarDiffusivity(; ν=1e-4, κ=1e-4)` doesn't allocate any fields, so no worries about that.

Do you mind sharing your script / model setup so we can look into it further?

**Reply by liuchihl** [reply]
Ah, you're right about that!

Don't mind at all, here is the [script](https://github.com/liuchihl/internal-tide-mixing/blob/023d28941eb7b815f3036a13d119f04fe7345142/test.jl). 
Thanks for your time!

**Reply by glwagner** [reply]
Nothing stands out at first glance. What is the discrepancy now?

Also some random comments:

* your above plot might be easier for me to interpret if the variable that is changed (the simulation size) is on the x-axis, and the measured variable (GPU memory consumption) is on the y-axis

* you're using no-slip boundary condition on the immersed boundary, but I don't think you resolve a molecular boundary layer... you may want to consider a boundary condition more appropriate for LES / WENO advection (a quadratic drag following law of the wall theory for example)

* since your grid is vertically stretched, the pressure solver is using `FourierTridiagonalPoissonSolver`: 

https://github.com/CliMA/Oceananigans.jl/blob/main/src/Solvers/fourier_tridiagonal_poisson_solver.jl

* I _think_ this solver just has one complex-valued scratch field:

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/fourier_tridiagonal_poisson_solver.jl#L82

**Reply by liuchihl** [reply]
The discrepancy is better than my first estimation now, but it still does not match the 1 to 1 line (assuming this is possible). I have interchanged the x and y axis as below. To make the estimation the same as the measured GPU usage, about 7 more variables are needed (i.e., N_state = 23).
![benchmark_mem_better_estimation_changecoordinate](https://github.com/CliMA/Oceananigans.jl/assets/68127124/97006d99-9ae5-491e-9f20-6db6bb13c63d)

> you're using no-slip boundary condition on the immersed boundary, but I don't think you resolve a molecular boundary layer... you may want to consider a boundary condition more appropriate for LES / WENO advection (a quadratic drag following law of the wall theory for example)

Thanks for the suggestion, I agree implementing a quadratic drag is a good idea since it would be pretty expensive to resolve the boundary layer, plus my goal is not to focus on boundary layer physics.  

> since your grid is vertically stretched, the pressure solver is using FourierTridiagonalPoissonSolver:

That's good to know! So adding this complex-valued scratch field makes `N_state=16`. There could be some scratch fields that I am not aware of. This might be a dumb question: is there a tip on identifying such scratch fields in the source codes? I do see the name is usually related to "storage".

**Reply by glwagner** [reply]
> This might be a dumb question: is there a tip on identifying such scratch fields in the source codes? I do see the name is usually related to "storage".

There's no convention. You might be able to figure out how to recursively search the properties of `NonhydrostaticModel` to list all array-like variables (ie variables with type `AbstractArray`, or that take up a significant amount of memory) though...

**Reply by glwagner** [reply]
Ok, I found four more fields: the `diagonal` of the tridiagonal solver:

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/fourier_tridiagonal_poisson_solver.jl#L91

Also, the tridiagonal solver itself builds a scratch field:

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/fourier_tridiagonal_poisson_solver.jl#L104

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/batched_tridiagonal_solver.jl#L69

Finally, the pressure solver also allocates memory for the solution (which has to be complex-valued to be compatible with the fourier transform, despite that we are typically looking for a real-valued pressure):

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/fourier_tridiagonal_poisson_solver.jl#L111

There can also be a buffer which would add again 2 more fields:

https://github.com/CliMA/Oceananigans.jl/blob/18c67c82d462612178f333ef6eb97b159d8373cd/src/Solvers/fourier_tridiagonal_poisson_solver.jl#L108

Perhaps you can check if there is a `model.pressure_solver.buffer` for your case.

**Reply by liuchihl** [reply]
Thanks for checking this! Looks like the estimation is getting really close to the actual usage!

**Reply by liuchihl** [reply]
> Perhaps you can check if there is a model.pressure_solver.buffer for your case.

When I run `julia> model.pressure_solver.buffer`, the output is empty, so I guess there is no buffer there.

**Reply by francispoulin** [reply]
All very interesting to see and I just have two thoughts.

1. What if you try saving just one field?  I know that's not as exicting but if you get there's a descrptency when you only save the buoyancy, then maybe that will shed some light on what happens with other fields.
2. Would you consider doing this with the (experimental) shallow water model? It does not involve any pressure solvers so that might be easier to sort out.

**Reply by liuchihl** [reply]
Thanks for your thoughts. 

>  What if you try saving just one field? I know that's not as exicting but if you get there's a descrptency when you only save the buoyancy, then maybe that will shed some light on what happens with other fields.

I agree, that's a good idea. I can maybe even run without saving any fields, excluding any memory usage for output writer (though small), the discrepancy would be smaller. 

> Would you consider doing this with the (experimental) shallow water model? It does not involve any pressure solvers so that might be easier to sort out.

That's a good point, it would be a great experiment to do.

**Reply by glwagner** [reply]
> I agree, that's a good idea. I can maybe even run without saving any fields, excluding any memory usage for output writer (though small), the discrepancy would be smaller.

This is also all an approximation because there are many 1D arrays that are allocated, eg for the stretched grid, the tridiagonal solver, etc.

**Reply by glwagner** [reply]
You might just remove the grid stretching to see how much usage changes. If you do that, you'll use a simpler pressure solver and the grid metrics will be constant rather than arrays. And the grid nodes will be ranges rather than arrays.

**Reply by liuchihl** [reply]
Thanks for the suggestions, I will try that out!
